### ISSUE TITLE
[Bug]: Asking for OpenAI key after loading index from disc.

### ISSUE THREAD

**Issue by AnuragSa** [run-llama/llama_index#8615]
### Bug Description

I am using AzureOpenAI embedding and llm which is set in ServiceContext. Once the index is generated, it is persisted to a local folder. While loading index from storage, it asks for OpenAI key when I have already configured it to use AzureOpenAI. 
The issue occurs on line 47. It works fine when the index is created for the first time.

<img width="1177" alt="Screenshot 2023-10-31 at 8 13 46 PM" src="https://github.com/run-llama/llama_index/assets/15018852/e414a839-f1c1-438e-b28e-116e4c88bf25">

### Version

0.8.56

### Steps to Reproduce

Run the file. Index is created fine using AzureOpenAi llm and embeddings. When run again, after loading index from storage, the issue occurs.

### Relevant Logs/Tracbacks

```shell
Error Log:
Traceback (most recent call last):
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/llms/utils.py", line 20, in resolve_llm
    llm = OpenAI()
          ^^^^^^^^
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/llms/openai.py", line 92, in __init__
    api_key, api_type, api_base, api_version = resolve_openai_credentials(
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/llms/openai_utils.py", line 297, in resolve_openai_credentials
    raise ValueError(MISSING_API_KEY_ERROR_MESSAGE)
ValueError: No API key found for OpenAI.
Please set either the OPENAI_API_KEY environment variable or openai.api_key prior to initialization.
API keys can be found or created at https://platform.openai.com/account/api-keys


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/anurag/Anurag Work/Coding/everything-ai/llmaindex-pdf/main.py", line 47, in <module>
    index = load_index_from_storage(storage_context)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/indices/loading.py", line 33, in load_index_from_storage
    indices = load_indices_from_storage(storage_context, index_ids=index_ids, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/indices/loading.py", line 78, in load_indices_from_storage
    index = index_cls(
            ^^^^^^^^^^
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/indices/vector_store/base.py", line 49, in __init__
    super().__init__(
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/indices/base.py", line 61, in __init__
    self._service_context = service_context or ServiceContext.from_defaults()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/indices/service_context.py", line 157, in from_defaults
    llm_predictor = llm_predictor or LLMPredictor(
                                     ^^^^^^^^^^^^^
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/llm_predictor/base.py", line 92, in __init__
    self._llm = resolve_llm(llm)
                ^^^^^^^^^^^^^^^^
  File "/Users/anurag/Anurag Work/Coding/everything-ai/.venv/lib/python3.11/site-packages/llama_index/llms/utils.py", line 22, in resolve_llm
    raise ValueError(
ValueError: ******
Could not load OpenAI model. If you intended to use OpenAI, please check your OPENAI_API_KEY.
Original error:
No API key found for OpenAI.
Please set either the OPENAI_API_KEY environment variable or openai.api_key prior to initialization.
API keys can be found or created at https://platform.openai.com/account/api-keys

To disable the LLM entirely, set llm=None.
******
```

**Comment by AnuragSa** [comment]
This dies not solve the issue for me. As I mentioned, this works first time but not when loading from storage.

**Comment by AnuragSa** [comment]
@dosu-bot, can you give me a code example of how to solve the issue?

**Comment by Disiok** [comment]
You need to recreate and pass in the service context while loading the index (i.e. when calling `load_index_from_storage`).

**Comment by scorpio0101** [comment]
Hi @Disiok,

Ran the following before loading the index from disk (still showing the same error mentioned above):

remotely_run_llm = HuggingFaceInferenceAPI(
    model_name="HuggingFaceH4/zephyr-7b-beta", token=HUGGINGFACEHUB_API_TOKEN
)

remote_embed_model = HuggingFaceInferenceAPIEmbedding(model_name="BAAI/bge-base-en-v1.5")

service_context = ServiceContext.from_defaults(
    chunk_size = 1024,
    llm = remotely_run_llm,
    embed_model = remote_embed_model
)

# rebuild storage context
storage_context = StorageContext.from_defaults(persist_dir='/content/drive/MyDrive/Colab Notebooks/storage')

# load index
index_from_storage = load_index_from_storage(storage_context)

**Comment by Disiok** [comment]
try `index_from_storage = load_index_from_storage(storage_context, service_context)`

**Comment by scorpio0101** [comment]
Thanks but please see below:

TypeError                                 Traceback (most recent call last)
[<ipython-input-12-8e9852e6b7ce>](https://localhost:8080/#) in <cell line: 12>()
     10 
     11 # load index
---> 12 index_from_storage = load_index_from_storage(storage_context, service_context)
     13 index_from_storage

3 frames
[/usr/local/lib/python3.10/dist-packages/llama_index/storage/kvstore/simple_kvstore.py](https://localhost:8080/#) in get(self, key, collection)
     38         if not collection_data:
     39             return None
---> 40         if key not in collection_data:
     41             return None
     42         return collection_data[key].copy()

TypeError: unhashable type: 'ServiceContext'

**Comment by AnuragSa** [comment]
I found out that adding the following line after creating a ServiceContext solves the problem. In my original screenshot, that would be line #39.

`set_global_service_context(service_context)`

https://gpt-index.readthedocs.io/en/v0.6.28/how_to/customization/service_context.html

**Comment by scorpio0101** [comment]
setting the service_context to be global worked. thanks.

I noticed that there is another issue when using the VectorStore:

index = VectorStoreIndex.from_documents(documents, service_context=service_context)

RuntimeError                              Traceback (most recent call last)
[<ipython-input-8-59f38e0d0f3e>](https://localhost:8080/#) in <cell line: 4>()
      2 documents = SimpleDirectoryReader('/content/drive/MyDrive/Colab Notebooks/data').load_data()
      3 # index = SummaryIndex.from_documents(documents, service_context=service_context)
----> 4 index = VectorStoreIndex.from_documents(documents, service_context=service_context)
      5 

11 frames
[/usr/lib/python3.10/asyncio/base_events.py](https://localhost:8080/#) in _check_running(self)
    584             raise RuntimeError('This event loop is already running')
    585         if events._get_running_loop() is not None:
--> 586             raise RuntimeError(
    587                 'Cannot run the event loop while another loop is running')
    588 

RuntimeError: Cannot run the event loop while another loop is running

**Comment by VArdulov** [comment]
Having looked into the details of the implementation @Disiok almost had the correct answer to this problem:

> try `index_from_storage = load_index_from_storage(storage_context, service_context)`

TL;DR: put the keyword in front of the service_context, because it's not a default parameter for the `load_index_from_storage`
i.e.:

```
index_from_storage = load_index_from_storage(
    storage_context=storage_context,
    service_context=service_context
)
```
this should load the index from disk without needing any OpenAI key, however notably make sure you set up the ServiceContext with the correct embedding model, otherwise you won't be able to look up things correctly. 

**Detailed Explanation:**
Looking specifically at the code for the `load_index_from_storage`, the `storage_context` and `index_id` are the two arguments you can pass in without needing the keywords passed in as well. So the issue with the code before was that since you weren't explicitly calling out that the service_context was the service_context the function assumed it was the index_id, which is not hashable, meaning you can't use it to look up the index in a dictionary, ergo the error @scorpio0101 was seeing. The code for `load_index_from_storage` actually invokes `load_indices_from_storage` with all of the same  kwargs but only returns the first result. Digging a little deeper it's clear that at one point the persisted index is loaded into a new instance of the index class, this is where the `**kwargs` can come be defined, otherwise the constructor will assume the defaults, which is why if you don't pass an explicit instance of the ServiceContext with the correct `embed_model` it will go looking for the `OPENAI_API_KEY` in your environment. This could be mitigated if in future versions `llama_index`, if during the index saving step the ServiceContext was saved as well, however there are a lot of issues with trying to seriously do this (allowing for custom index types, and custom LLMs makes this a tough). My recommendation if you are seriously trying to persist this, is to save a JSON that has all of the information regarding the embedding model and other information so that you can pass those in programmatically in the future and save it as a log.

**Comment by scorpio0101** [comment]
Thanks for the reply. 
The motivation behind saving the index to disk is mainly to save compute resources for every subsequent lookup on the same document. It would be a waste of resources to embed the text every time I run the code. 

Two more issues I noticed, the combination of remote embedding and Vectorstore do not seem to work. 

**Error**:
RuntimeError Traceback (most recent call last)
in <cell line: 4>()
2 documents = SimpleDirectoryReader('/content/drive/MyDrive/Colab Notebooks/data').load_data()
3 # index = SummaryIndex.from_documents(documents, service_context=service_context)
----> 4 index = VectorStoreIndex.from_documents(documents, service_context=service_context)
5

11 frames
[/usr/lib/python3.10/asyncio/base_events.py](https://localhost:8080/#) in _check_running(self)
584 raise RuntimeError('This event loop is already running')
585 if events._get_running_loop() is not None:
--> 586 raise RuntimeError(
587 'Cannot run the event loop while another loop is running')
588

RuntimeError: Cannot run the event loop while another loop is running


Also, for some reason, the query engine no longer seem to have access to the vectorstore.
I checked both documents and vector store do have the relevant contents. 

Strange behavior. Was working well with the same code a couple of days ago.

**Comment by logan-markewich** [comment]
That error just means you have to enable nesting async

```
import nest_asyncio

nest_asyncio.apply()
```

Going to close this issue for now :)

**Comment by VArdulov** [comment]
@scorpio0101 I don't know understand what you mean by this:
> saving the index to disk is mainly to save compute resources for every subsequent lookup on the same document. It would be a waste of resources to embed the text every time I run the code.

The solution I provided doesn't re-index the documents each time. It just initializes the `ServiceContext`, i.e. you are simply giving your environment the necessary information to look use the appropriate LLM/Embedding models for interfacing with your index which you have loaded from the disk. The `VectorStoreIndex` converts "documents" to vector representation according to the embedding model chosen when the index is constructed (by default it uses OpenAI's GPT model, which is why if you don't initialize `ServiceContext` it will look for the API keys in your environment). Since that information isn't stored in anyway when you persist your index to the disk, you need to recreate it when you load the index from the disk.

**Comment by scorpio0101** [comment]
Thanks @VArdulov.

**Comment by hristochr** [comment]
The [first reply](https://github.com/run-llama/llama_index/issues/8615#issuecomment-1788222225) in fact helped me solve an identical issue. To bypass the error, I initialize my embedding model, although I don't need to use it anymore. Then I set the global `embed_model` setting:
![image](https://github.com/run-llama/llama_index/assets/22388580/19ae59e2-49a6-45ba-b6d0-5b37203b2ab7)
